server.go

```diff
@@ -1868,16 +1850,47 @@ func (s *Server) GracefulStop() {
 	defer s.done.Fire()
 
 	s.channelzRemoveOnce.Do(func() { channelz.RemoveEntry(s.channelzID) })
+
 	s.mu.Lock()
-	if s.conns == nil {
-		s.mu.Unlock()
-		return
+	s.closeListeners()
+	s.drainAllServerTransports()
+	s.mu.Unlock()
+
+	// Wait for serving threads to be ready to exit.  Only then can we be sure no
+	// new conns will be created.
+	s.serveWG.Wait()
```

Comment:
> Don't we need to do this before we drain the transports, then?

The order looks correct to me:
1. First we close all listeners to prevent that new TCP connections are accepted
2. Then we call `drainAllServerTransports`, inform the clients about the incoming shutdown by sending GOAWAY
3. Then wait until all connections were closed (`s.serveWG.Wait()`).

Does that make sense or am I missing?


> Can we unify them even more, and have a stop(drain bool) or stop(stopTransport(*transport.ServerTransport)) method that the existing GracefulStop() and Stop() call to do everything?

great idea, I'll have a look

Reply:
changed

Reply:
>> Don't we need to do this before we drain the transports, then?
> The order looks correct to me:
> 1. First we close all listeners to prevent that new TCP connections are accepted
> 2. Then we call drainAllServerTransports, inform the clients about the incoming shutdown by sending GOAWAY

I think there's a misunderstanding on your part.  `serveWG.Wait()` unblocks when all `Serve` calls unblock.  This happens when the listeners are successfully closed and we're done accepting new connections.

Your call to `drain` is before the call to `serveWG.Wait()`, but if we don't do `serveWG.Wait()` first, then we might miss draining a new transport.

Reply:
thanks, got it!

I changed the order